### PR TITLE
Adapt the new exception system (#302) for master

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | maybe
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #302 
| License       | MIT
| Doc PR        | n/a

Adapt #302 to master.

### CODE REVIEW

File: Tests/DependencyInjection/DunglasApiExtensionTest.php [hunk 1]
```diff
@@ -192,6 +192,7 @@ private function getContainerBuilderProphecy()
         $containerBuilderProphecy->setDefinition('api.jsonld.action.context', $definitionArgument)->shouldBeCalled();
         $containerBuilderProphecy->setDefinition('api.jsonld.action.entrypoint', $definitionArgument)->shouldBeCalled();
         $containerBuilderProphecy->setDefinition('api.hydra.documentation_builder', $definitionArgument)->shouldBeCalled();
+	$containerBuilderProphecy->setDefinition('api.hydra.controller.exception', $definitionArgument)->shouldBeCalled();
```
Comment (sroze) [inline Tests/DependencyInjection/DunglasApiExtensionTest.php]: Please replace this tab by 8 spaces :wink:

Comment (dunglas) [inline Tests/DependencyInjection/DunglasApiExtensionTest.php]: The vi effect

**Comment by sroze** [comment]
Can't this be directly done by merging 1.0 in master?

**Comment by dunglas** [comment]
No because of the ADR pattern (and there is a missing file in this PR, I'm trying to recover it after a laptop crash).

**Comment by dunglas** [comment]
ping @meyerbaptiste @ogizanagi

File: Hydra/EventListener/RequestExceptionListener.php [hunk 2]
```diff
@@ -25,52 +20,19 @@
...
-
     /**
      * @param GetResponseForExceptionEvent $event
      */
     public function onKernelException(GetResponseForExceptionEvent $event)
     {
+        // Normalize exceptions with hydra errors only for resources
         $request = $event->getRequest();
-        if (!$request->attributes->has('_resource_type') || self::FORMAT !== $request->attributes->get('_api_format')) {
+        if (!$request->attributes->has('_resource_type')) {
```
Comment (ogizanagi) [inline Hydra/EventListener/RequestExceptionListener.php]: `$event->getRequest()` could be inlined here IMO (as it is in 1.0), as it is the unique usage here and is still pretty readable.

File: Hydra/Action/ExceptionAction.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+     *
+     * @param FlattenException $exception
+     *
+     * @return Response
+     */
+    public function __invoke(FlattenException $exception)
+    {
+        $exceptionClass = $exception->getClass();
+        if (
+            is_a($exceptionClass, DeserializationException::class, true) ||
```
Comment (ogizanagi) [inline Hydra/Action/ExceptionAction.php]: `DeserializationException` implements `Symfony\Component\Serializer\Exception\ExceptionInterface`

File: Hydra/Serializer/ErrorNormalizer.php [hunk 4]
```diff
@@ -11,19 +11,19 @@
...
- * Converts {@see \Exception} to a Hydra error representation.
+ * Converts {@see \Exception} or {@see \Symfony\Component\Debug\Exception\FlattenException}
+ * to a Hydra error representation.
  *
  * @author Kévin Dunglas <dunglas@gmail.com>
  * @author Samuel ROZE <samuel.roze@gmail.com>
  */
 class ErrorNormalizer implements NormalizerInterface
 {
-    const FORMAT = 'hydra-error';
```
Comment (ogizanagi) [inline Hydra/Serializer/ErrorNormalizer.php]: Why removing this constant ?

File: Hydra/EventListener/ValidationExceptionListener.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+     * Returns a list of violations normalized in the Hydra format.
+     *
+     * @param GetResponseForExceptionEvent $event
+     */
+    public function onKernelException(GetResponseForExceptionEvent $event)
+    {
+        $exception = $event->getException();
+
+        if ($exception instanceof ValidationException) {
+            $event->setResponse(new Response(
```
Comment (meyerbaptiste) [inline Hydra/EventListener/ValidationExceptionListener.php]: In this case the propagation is stopped and #302 is no longer fixed!

Comment (dunglas) [inline Hydra/EventListener/ValidationExceptionListener.php]: It's intended. IMHO validation exception is special and should not be logged (as when using the form component in Symfony).

**Comment by dunglas** [comment]
@ogizanagi comments fixed.

File: Hydra/EventListener/ValidationExceptionListener.php (hunk 5 above)
Comment (ogizanagi) [inline Hydra/EventListener/ValidationExceptionListener.php]: Agreed.

**Comment by ogizanagi** [comment]
:+1:

**Pull request merged**